contracts/modules/token/TokenController.sol

```diff
@@ -418,300 +161,235 @@ contract TokenController is ITokenController, LockHandler, LegacyMasterAware {
       }
 
-      lockReason[_of].pop();
-      emit Unlocked(_of, _reason, amount);
+      lockReason[user].pop();
+      emit Unlocked(user, _reason, amount);
 
       if (lastReasonIndex > 0) {
-        lastReasonIndex = lastReasonIndex.sub(1, "TokenController: Reason count mismatch");
+        lastReasonIndex = lastReasonIndex - 1;
       }
     }
 
-    token.transfer(_of, totalAmount);
+    token.transfer(user, totalAmount);
   }
 
-  function removeEmptyReason(address _of, bytes32 _reason, uint _index) external {
-    _removeEmptyReason(_of, _reason, _index);
-  }
-
-  function removeMultipleEmptyReasons(
-    address[] calldata _members,
-    bytes32[] calldata _reasons,
-    uint[] calldata _indexes
-  ) external {
 
-    require(_members.length == _reasons.length, "TokenController: members and reasons array lengths differ");
-    require(_reasons.length == _indexes.length, "TokenController: reasons and indexes array lengths differ");
+  function mintStakingPoolNXMRewards(uint amount, uint poolId) external {
 
-    for (uint i = _members.length; i > 0; i--) {
-      uint idx = i - 1;
-      _removeEmptyReason(_members[idx], _reasons[idx], _indexes[idx]);
-    }
+    require(msg.sender == address(cover), "TokenController: only Cover allowed");
+    mint(address(this), amount);
+    stakingPoolNXMBalances[poolId].rewards += amount.toUint128();
   }
 
-  function _removeEmptyReason(address _of, bytes32 _reason, uint _index) internal {
+  function burnStakingPoolNXMRewards(uint amount, uint poolId) external {
 
-    uint lastReasonIndex = lockReason[_of].length.sub(1, "TokenController: lockReason is empty");
+    require(msg.sender == address(cover), "TokenController: only Cover allowed");
+    burnFrom(address(this), amount);
+    stakingPoolNXMBalances[poolId].rewards -= amount.toUint128();
+  }
 
-    require(lockReason[_of][_index] == _reason, "TokenController: bad reason index");
-    require(locked[_of][_reason].amount == 0, "TokenController: reason amount is not zero");
+  function depositStakedNXM(address from, uint amount, uint poolId) external {
+    require(msg.sender == address(cover.stakingPool(poolId)), "TokenController: msg.sender not staking pool");
 
-    if (lastReasonIndex != _index) {
-      lockReason[_of][_index] = lockReason[_of][lastReasonIndex];
-    }
+    stakingPoolNXMBalances[poolId].deposits += amount.toUint128();
+    token.operatorTransfer(from, amount);
+  }
 
-    lockReason[_of].pop();
+  function withdrawNXMStakeAndRewards(address to, uint stakeToWithdraw, uint rewardsToWithdraw, uint poolId) external {
+    require(msg.sender == address(cover.stakingPool(poolId)), "TokenController: msg.sender not staking pool");
+
+    StakingPoolNXMBalances memory currentBalances = stakingPoolNXMBalances[poolId];
+    stakingPoolNXMBalances[poolId] = StakingPoolNXMBalances(
+      currentBalances.deposits - stakeToWithdraw.toUint128(),
+      currentBalances.rewards - rewardsToWithdraw.toUint128()
+    );
```

Comment:
The first parameter in the `StakingPoolNXMBalances` is the `rewards` field, not the `deposits`!

Use something like this instead:

```
    StakingPoolNXMBalances memory poolBalances = stakingPoolNXMBalances[poolId];
    poolBalances.deposits -= stakeToWithdraw.toUint128();
    poolBalances.rewards -= rewardsToWithdraw.toUint128();
    stakingPoolNXMBalances[poolId] = poolBalances;
```